File: global-hooks/framework/guardrails/orchestrator_tool_guard.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.8"
# dependencies = []
# ///
"""
Orchestrator Tool Guard - PreToolUse Hook
==========================================

Prevents the orchestrator agent from directly using Read, Grep, Glob, Edit,
or Bash tools. The orchestrator's role is to delegate ALL file operations
to subagents (researchers, builders, validators).

Detection mechanism:
  - enforce_orchestrate.py (UserPromptSubmit) writes a marker file when
    /orchestrate is detected: /tmp/caf_orch_guard.marker
  - This hook checks for the marker AND whether the current call is at
    the orchestrator depth (depth 1 = orchestrator, depth 2+ = subagents)
  - SubagentStart hook increments depth, SubagentStop decrements it
  - Only blocks at depth 1 (orchestrator itself). Depth 0 (main session)
    and depth 2+ (subagents) are always allowed.

Trigger: PreToolUse (matcher: Read|Grep|Glob|Edit|Bash)
Exit codes:
  0 = Allow (not in orchestration mode, or subagent at depth 2+)
  2 = Block (orchestrator attempting forbidden tool)
"""

import json
import os
import sys
from pathlib import Path

# Marker file written by enforce_orchestrate.py
MARKER_FILE = Path("/tmp/caf_orch_guard.marker")

# Depth tracker: incremented by SubagentStart, decremented by SubagentStop
DEPTH_FILE = Path("/tmp/caf_orch_depth")

# Tools the orchestrator must never call directly
FORBIDDEN_TOOLS = {"Read", "Grep", "Glob", "Edit"}

# Bash is partially restricted — only cat/grep/find/head/tail/sed/awk
BASH_RESEARCH_PATTERNS = [
    "cat ", "grep ", "rg ", "find ", "head ", "tail ", "sed ", "awk ",
    "less ", "more ", "wc ", "sort ", "uniq ", "xargs ",
    "ls ", "tree ", "file ", "stat ",
]

# Block message injected via JSON output (additionalContext)
BLOCK_REMINDER = (
    "[ORCHESTRATOR GUARD] You are the orchestrator. You MUST NOT use "
    "file-reading or code-searching tools directly. Spawn a subagent instead:\n"
    "  - Need to read a file? → Agent(name='researcher-N', model='haiku', ...)\n"
    "  - Need to search code? → Agent(name='researcher-N', model='sonnet', ...)\n"
    "  - Need to edit code? → Agent(name='builder-N', model='sonnet', ...)\n"
    "  - Need to run tests? → Agent(name='validator-N', model='haiku', ...)\n\n"
    "This tool call has been BLOCKED. Delegate this work to a subagent."
)


def get_depth() -> int:
    """Read the current orchestration depth from the depth file."""
    try:
        if DEPTH_FILE.exists():
            return int(DEPTH_FILE.read_text().strip())
    except (ValueError, OSError):
        pass
    return 0


def is_bash_research(command: str) -> bool:
    """Check if a bash command is doing file research (cat, grep, etc.)."""
    cmd = command.strip().lstrip("'\"")
    return any(cmd.startswith(p) for p in BASH_RESEARCH_PATTERNS)


def main():
    try:
        # ── CAF_MODE check ────────────────────────────────────────────
        _framework_dir = Path(__file__).parent.parent
        if str(_framework_dir) not in sys.path:
            sys.path.insert(0, str(_framework_dir))
        try:
            from caf_mode import should_run
            if not should_run("orchestrator_tool_guard"):
                sys.exit(0)
        except ImportError:
            pass

        # ── Quick exit: no orchestration active ───────────────────────
        if not MARKER_FILE.exists():
            sys.exit(0)

        # ── Check depth: only block at depth 1 (orchestrator itself)
        depth = get_depth()
        if depth == 0:
            # This is the main session — never block it
            sys.exit(0)
        if depth >= 2:
            # This is a subagent's subagent (researcher, builder, etc.)
            # They SHOULD use Read/Grep/Glob/Edit — allow
            sys.exit(0)

        # ── Read hook input ───────────────────────────────────────────
        input_data = json.load(sys.stdin)
        tool_name = input_data.get("tool_name", "")
        tool_input = input_data.get("tool_input", {})

        # ── Check if tool is forbidden ────────────────────────────────
        should_block = False

        if tool_name in FORBIDDEN_TOOLS:
            should_block = True
        elif tool_name == "Bash":
            command = tool_input.get("command", "") if isinstance(tool_input, dict) else ""
            if is_bash_research(command):
                should_block = True

        if should_block:
            # Output block decision + reminder context
            print(BLOCK_REMINDER, file=sys.stderr)
            sys.exit(2)

        # Allow all other tools (Agent, Task, Write, SendMessage, etc.)
        sys.exit(0)

    except Exception:
        # Never crash, never block on error — fail open
        sys.exit(0)


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass

File: caf-hooks/src/hooks/orchestrator_tool_guard.rs
/// Orchestrator Tool Guard — PreToolUse hook.
///
/// Python equivalent: global-hooks/framework/guardrails/orchestrator_tool_guard.py (137 LOC)
///
/// Behavior:
/// - Check guard.marker exists at ~/.caf/orch_state/guard.marker
/// - Check depth file for current depth (default 0)
/// - Only block at depth == 1 (the orchestrator itself)
/// - If marker exists AND depth == 1 AND tool is Read/Grep/Glob/Edit → exit 2 (block)
/// - If marker exists AND depth == 1 AND tool is Bash AND command starts with
///   research patterns → exit 2 (block)
/// - Otherwise → print {} to stdout and exit 0 (fail-open)
use serde_json::Value;
use std::process;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::io::read_stdin_value;
use crate::state::{orch_depth_path, orch_guard_marker_path};

const STALE_ORCH_STATE_SECS: u64 = 86400;

/// Tools the orchestrator must never call directly (besides Bash research).
const FORBIDDEN_TOOLS: &[&str] = &["Read", "Grep", "Glob", "Edit", "NotebookEdit", "Write"];

/// Bash command prefixes that indicate file research (should be delegated).
const BASH_RESEARCH_PATTERNS: &[&str] = &[
    "cat ", "grep ", "rg ", "find ", "head ", "tail ", "sed ", "awk ",
    "less ", "more ", "wc ", "sort ", "uniq ", "xargs ",
    "ls ", "tree ", "file ", "stat ",
];

const BLOCK_REMINDER: &str = "\
[ORCHESTRATOR GUARD] You are the orchestrator. You MUST NOT use \
file-reading or code-searching tools directly. Spawn a subagent instead:\n\
  - Need to read a file? \u{2192} Agent(name='researcher-N', model='haiku', ...)\n\
  - Need to search code? \u{2192} Agent(name='researcher-N', model='sonnet', ...)\n\
  - Need to edit code? \u{2192} Agent(name='builder-N', model='sonnet', ...)\n\
  - Need to run tests? \u{2192} Agent(name='validator-N', model='haiku', ...)\n\n\
This tool call has been BLOCKED. Delegate this work to a subagent.";

fn get_depth() -> i64 {
    let path = orch_depth_path();
    if !path.exists() {
        return 0;
    }
    match std::fs::read_to_string(&path) {
        Ok(s) => {
            let s = s.trim();
            // Try JSON format first: {"depth": N, "ts": "..."}
            if let Ok(v) = serde_json::from_str::<Value>(s) {
                if let Some(d) = v.get("depth").and_then(|d| d.as_i64()) {
                    return d;
                }
            }
            // Fall back to raw integer (old format)
            s.parse::<i64>().unwrap_or(0)
        }
        Err(_) => 0,
    }
}

/// Read marker file JSON, parse "ts" field, check if age > STALE_ORCH_STATE_SECS.
/// Returns true if marker is stale (should be treated as absent).
fn is_marker_stale() -> bool {
    let path = orch_guard_marker_path();
    let contents = match std::fs::read_to_string(&path) {
        Ok(c) => c,
        Err(_) => return false,
    };
    let v: Value = match serde_json::from_str(contents.trim()) {
        Ok(v) => v,
        Err(_) => return false,
    };
    let ts_str = match v.get("ts").and_then(|t| t.as_str()) {
        Some(s) => s,
        None => return false,
    };
    // Parse RFC3339 timestamp using chrono
    let dt = match chrono::DateTime::parse_from_rfc3339(ts_str) {
        Ok(d) => d,
        Err(_) => return false,
    };
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    let marker_secs = dt.timestamp() as u64;
    let age = now.saturating_sub(marker_secs);
    if age > STALE_ORCH_STATE_SECS {
        let age_hours = age / 3600;
        eprintln!("[CAF] Ignoring stale orchestration guard marker (age: {}h)", age_hours);
        return true;
    }
    false
}

fn is_bash_research(command: &str) -> bool {
    let cmd = command.trim().trim_start_matches(|c| c == '\'' || c == '"');
    BASH_RESEARCH_PATTERNS.iter().any(|p| cmd.starts_with(p))
}

pub fn run() {
    // Fail-open wrapper — any internal error → exit 0
    if let Err(_) = run_inner() {
        println!("{{}}");
        process::exit(0);
    }
}

fn run_inner() -> Result<(), Box<dyn std::error::Error>> {
    // Quick exit: no orchestration active
    if !orch_guard_marker_path().exists() {
        println!("{{}}");
        return Ok(());
    }

    // Check if marker is stale — if so, treat as absent
    if is_marker_stale() {
        println!("{{}}");
        return Ok(());
    }

    // Only block at depth 1 (orchestrator itself).
    // Depth 0 = main session (never block), depth 2+ = subagent (allowed).
    let depth = get_depth();
    if depth != 1 {
        println!("{{}}");
        return Ok(());
    }

    let data: Value = read_stdin_value();
    let tool_name = data.get("tool_name").and_then(|v| v.as_str()).unwrap_or("");
    let tool_input = data.get("tool_input").cloned().unwrap_or(Value::Object(Default::default()));

    let should_block = if FORBIDDEN_TOOLS.contains(&tool_name) {
        true
    } else if tool_name == "Bash" {
        let command = tool_input
            .get("command")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        is_bash_research(command)
    } else {
        false
    };

    if should_block {
        eprintln!("{}", BLOCK_REMINDER);
        process::exit(2);
    }

    println!("{{}}");
    Ok(())
}
